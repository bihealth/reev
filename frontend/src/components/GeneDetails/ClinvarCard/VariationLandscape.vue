<script setup lang="ts">
import { computed } from 'vue'

import VegaPlot from '@/components/VegaPlot.vue'
import type { GenomeBuild } from '@/lib/genomeBuilds'

export interface Props {
  /** Gene information from annonars. */
  clinvar: any
  /** Transctipts information. */
  transcripts: any
  /** The genome release. */
  genomeBuild?: GenomeBuild
  /** Gene symbol */
  geneSymbol: string
}

const props = withDefaults(defineProps<Props>(), {
  clinvar: null,
  transcripts: null,
  genomeBuild: 'grch37',
  geneSymbol: '???'
})

const clinvarSignificanceMapping: { [key: string]: number } = {
  CLINICAL_SIGNIFICANCE_UNKNOWN: -3,
  CLINICAL_SIGNIFICANCE_PATHOGENIC: 2,
  CLINICAL_SIGNIFICANCE_LIKELY_PATHOGENIC: 1,
  CLINICAL_SIGNIFICANCE_UNCERTAIN_SIGNIFICANCE: 0,
  CLINICAL_SIGNIFICANCE_LIKELY_BENIGN: -1,
  CLINICAL_SIGNIFICANCE_BENIGN: -2
}

interface ClinvarReferenceAssertions {
  rcv: string
  clinicalSignificance: string
  reviewStatus: string
  title: string
}

interface ClinvarVariant {
  chrom: string
  start: number
  stop: number
  reference: string
  alternative: string
  vcv: string
  referenceAssertions: ClinvarReferenceAssertions[]
}

const convertClinvarSignificance = (input: string): number => {
  if (input in clinvarSignificanceMapping) {
    return clinvarSignificanceMapping[input]
  } else {
    return -4
  }
}

const minMax = computed(() => {
  if (!props.clinvar) {
    return []
  }
  let min = null
  let max = null
  for (const item of props.clinvar.variants ?? []) {
    if (item.genomeRelease.toLowerCase() == props.genomeBuild) {
      // Go through all item.variants and find the min and max pos.
      for (const variant of item.variants) {
        if (variant.start < min || min === null) {
          min = variant.start
        }
        if (variant.start > max || max === null) {
          max = variant.start
        }
      }
    }
  }
  for (const exon of exons.value) {
    if (exon.start < min || min === null) {
      min = exon.start
    }
    if (exon.stop > max || max === null) {
      max = exon.stop
    }
  }
  const totalLength = max - min
  const padding = Math.round(totalLength * 0.05)
  return [min - padding, max + padding]
})

const paddedMinMax = computed(() => {
  const [min, max] = minMax.value
  const totalLength = max - min
  const padding = Math.round(totalLength * 0.05)
  return [min - padding, max + padding]
})

const exons = computed(() => {
  if (!props.transcripts?.length) {
    return []
  }
  const exons = []
  for (const transcript of props.transcripts.transcripts) {
    for (const alignment of transcript.alignments) {
      for (const exon of alignment.exons) {
        exons.push({
          start: exon.ref_start,
          stop: exon.ref_end
        })
      }
    }
  }
  return exons
})

const vegaData = computed(() => {
  if (!props.clinvar) {
    return []
  }
  let clinvarInfo = []
  for (const item of props.clinvar.variants ?? []) {
    if (item.genomeRelease.toLowerCase() == props.genomeBuild) {
      clinvarInfo = item.variants
    }
  }

  return clinvarInfo.map((variant: ClinvarVariant) => ({
    pos: variant.start,
    clinsig: convertClinvarSignificance(variant.referenceAssertions[0].clinicalSignificance)
  }))
})

const vegaEncoding = {}

const vegaLayer = [
  {
    description: 'gray baseline',
    data: { values: [{}] },
    mark: { type: 'rule', stroke: 'lightgray', size: 3 },
    encoding: { y: { datum: 'Uncertain significance' } }
  },
  {
    description: 'lollipop heads',
    transform: [
      {
        lookup: 'clinsig',
        from: {
          data: {
            values: [
              { clinsig: -5, clinsigLabel: 'other' },
              { clinsig: -4, clinsigLabel: 'Not provided' },
              { clinsig: -3, clinsigLabel: 'Conflicting' },
              { clinsig: -2, clinsigLabel: 'Benign' },
              { clinsig: -1, clinsigLabel: 'Likely benign' },
              { clinsig: 0, clinsigLabel: 'Uncertain significance' },
              { clinsig: 1, clinsigLabel: 'Likely pathogenic' },
              { clinsig: 2, clinsigLabel: 'Pathogenic' },
              { clinsig: 3, clinsigLabel: 'Gene' }
            ]
          },
          key: 'clinsig',
          fields: ['clinsigLabel']
        }
      }
    ],
    mark: { type: 'circle', opacity: 0.8 },
    encoding: {
      x: {
        field: 'pos',
        type: 'quantitative',
        scale: { domain: paddedMinMax.value },
        axis: { grid: false, zindex: 1000 },
        title: null
      },
      y: {
        field: 'clinsigLabel',
        type: 'nominal',
        scale: {
          domain: [
            'Gene',
            'Pathogenic',
            'Likely pathogenic',
            'Uncertain significance',
            'Likely benign',
            'Benign'
          ]
        },
        axis: { grid: false },
        title: null
      },
      color: {
        field: 'clinsigLabel',
        type: 'nominal',
        scale: {
          domain: [
            'Gene',
            'Pathogenic',
            'Likely pathogenic',
            'Uncertain significance',
            'Likely benign',
            'Benign'
          ],
          range: ['gray', 'darkred', 'orange', 'yellow', 'green', 'darkgreen']
        },
        legend: null
      },
      size: { value: 100 }
    }
  },
  {
    description: 'lollipop sticks',
    transform: [
      {
        lookup: 'clinsig',
        from: {
          data: {
            values: [
              {
                clinsig: -5,
                clinsigBaseline: 'Uncertain significance',
                clinsigLabel: 'other'
              },
              {
                clinsig: -4,
                clinsigBaseline: 'Uncertain significance',
                clinsigLabel: 'Not provided'
              },
              {
                clinsig: -3,
                clinsigBaseline: 'Uncertain significance',
                clinsigLabel: 'Conflicting'
              },
              {
                clinsig: -2,
                clinsigBaseline: 'Uncertain significance',
                clinsigLabel: 'Benign'
              },
              {
                clinsig: -1,
                clinsigBaseline: 'Uncertain significance',
                clinsigLabel: 'Likely benign'
              },
              {
                clinsig: 0,
                clinsigBaseline: 'Uncertain significance',
                clinsigLabel: 'Uncertain significance'
              },
              {
                clinsig: 1,
                clinsigBaseline: 'Uncertain significance',
                clinsigLabel: 'Likely pathogenic'
              },
              {
                clinsig: 2,
                clinsigBaseline: 'Uncertain significance',
                clinsigLabel: 'Pathogenic'
              },
              {
                clinsig: 3,
                clinsigBaseline: 'Uncertain significance',
                clinsigLabel: 'Gene'
              }
            ]
          },
          key: 'clinsig',
          fields: ['clinsigBaseline', 'clinsigLabel']
        }
      }
    ],
    mark: { type: 'rule', opacity: 0.8 },
    encoding: {
      x: {
        field: 'pos',
        type: 'quantitative',
        scale: { domain: paddedMinMax.value },
        axis: { grid: false },
        title: null
      },
      y: {
        field: 'clinsigLabel',
        type: 'nominal',
        scale: {
          domain: [
            'Gene',
            'Pathogenic',
            'Likely pathogenic',
            'Uncertain significance',
            'Likely benign',
            'Benign'
          ]
        },
        axis: { grid: false, zindex: 1000 },
        title: null
      },
      y2: { field: 'clinsigBaseline' },
      color: {
        field: 'clinsigLabel',
        type: 'nominal',
        scale: {
          domain: [
            'Gene',
            'Pathogenic',
            'Likely pathogenic',
            'Uncertain significance',
            'Likely benign',
            'Benign'
          ],
          range: ['gray', 'darkred', 'orange', 'yellow', 'green', 'darkgreen']
        },
        legend: null
      },
      size: { value: 1 }
    }
  },
  {
    description: 'gene - line',
    data: { values: [{ pos: minMax.value[0] }, { pos: minMax.value[1] }] },
    mark: { type: 'line', stroke: 'black', size: 1, opacity: 0.5 },
    encoding: {
      x: {
        field: 'pos',
        type: 'quantitative',
        scale: { domain: paddedMinMax.value },
        axis: { grid: false },
        title: null
      },
      y: { datum: 'Gene' }
    }
  },
  {
    description: 'gene - exons',
    data: {
      values: exons.value
    },
    mark: {
      type: 'rect',
      stroke: 'black',
      height: 10,
      fill: 'black',
      opacity: 0.5
    },
    encoding: {
      x: {
        field: 'start',
        type: 'quantitative',
        scale: { domain: paddedMinMax.value },
        axis: { grid: false },
        title: null
      },
      x2: { field: 'stop' },
      y: { datum: 'Gene' }
    }
  }
]
</script>

<template>
  <div class="pt-3">
    <template v-if="!clinvar">
      <v-skeleton-loader class="mt-3 mx-auto border" type="text,image,text" />
    </template>
    <v-sheet v-else rounded="lg" class="bg-grey-lighten-3 pa-3 mt-3 h-100">
      <div class="text-subtitle-1 text-center">Variation Lanscape</div>
      <VegaPlot
        :data-values="vegaData"
        :encoding="vegaEncoding"
        :layer="vegaLayer"
        :width="1200"
        :height="300"
        renderer="canvas"
      />
      <div>
        The plot above shows the sequence variants in the ClinVar database that are that are located
        in the gene <span class="font-italic"> {{ geneSymbol }} </span>.
      </div>
    </v-sheet>
  </div>
</template>
