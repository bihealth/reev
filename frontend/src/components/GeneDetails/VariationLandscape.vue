<script setup lang="ts">
import { computed } from 'vue'

import VegaPlot from '@/components/VegaPlot.vue'

export interface Props {
  /** Gene information from annonars. */
  clinvar: any
  /** Transctipts information. */
  transcripts: any
  /** The genome release. */
  genomeRelease: string
  /** The gene HGNC symbol. */
  hgnc: string
}

const props = withDefaults(defineProps<Props>(), {
  clinvar: null,
  transcripts: null,
  genomeRelease: 'grch37',
  hgnc: ''
})

const clinvarSignificanceMapping: { [key: string]: number } = {
  CLINICAL_SIGNIFICANCE_UNKNOWN: -3,
  CLINICAL_SIGNIFICANCE_PATHOGENIC: 2,
  CLINICAL_SIGNIFICANCE_LIKELY_PATHOGENIC: 1,
  CLINICAL_SIGNIFICANCE_UNCERTAIN_SIGNIFICANCE: 0,
  CLINICAL_SIGNIFICANCE_LIKELY_BENIGN: -1,
  CLINICAL_SIGNIFICANCE_BENIGN: -2
}

interface ClinvarReferenceAssertions {
  rcv: string
  clinicalSignificance: string
  reviewStatus: string
  title: string
}

interface ClinvarVariant {
  chrom: string
  start: number
  stop: number
  reference: string
  alternative: string
  vcv: string
  referenceAssertions: ClinvarReferenceAssertions[]
}

const convertClinvarSignificance = (input: string): number => {
  if (input in clinvarSignificanceMapping) {
    return clinvarSignificanceMapping[input]
  } else {
    return -4
  }
}

const minMax = computed(() => {
  if (!props.clinvar) {
    return []
  }
  let min = null
  let max = null
  for (const item of props.clinvar.variants ?? []) {
    if (item.genomeRelease.toLowerCase() == props.genomeRelease) {
      // Go through all item.variants and find the min and max pos.
      for (const variant of item.variants) {
        if (variant.start < min || min === null) {
          min = variant.start
        }
        if (variant.start > max || max === null) {
          max = variant.start
        }
      }
    }
  }
  for (const exon of exons.value) {
    if (exon.start < min || min === null) {
      min = exon.start
    }
    if (exon.stop > max || max === null) {
      max = exon.stop
    }
  }
  const totalLength = max - min
  const padding = Math.round(totalLength * 0.05)
  return [min - padding, max + padding]
})

const paddedMinMax = computed(() => {
  let [min, max] = minMax.value
  const totalLength = max - min
  const padding = Math.round(totalLength * 0.05)
  return [min - padding, max + padding]
})

const exons = computed(() => {
  if (!props.transcripts) {
    return []
  }
  const exons = []
  for (const transcript of props.transcripts.transcripts) {
    for (const alignment of transcript.alignments) {
      for (const exon of alignment.exons) {
        exons.push({
          start: exon.ref_start,
          stop: exon.ref_end
        })
      }
    }
  }
  return exons
})

const vegaData = computed(() => {
  if (!props.clinvar) {
    return []
  }
  let clinvarInfo = []
  for (const item of props.clinvar.variants ?? []) {
    if (item.genomeRelease.toLowerCase() == props.genomeRelease) {
      clinvarInfo = item.variants
    }
  }

  return clinvarInfo.map((variant: ClinvarVariant) => ({
    pos: variant.start,
    clinsig: convertClinvarSignificance(variant.referenceAssertions[0].clinicalSignificance)
  }))
})

const vegaEncoding = {}

const vegaLayer = [
  {
    description: 'gray baseline',
    data: { values: [{}] },
    mark: { type: 'rule', stroke: 'lightgray', size: 3 },
    encoding: { y: { datum: 'Uncertain significance' } }
  },
  {
    description: 'lollipop heads',
    transform: [
      {
        lookup: 'clinsig',
        from: {
          data: {
            values: [
              { clinsig: -5, clinsigLabel: 'other' },
              { clinsig: -4, clinsigLabel: 'Not provided' },
              { clinsig: -3, clinsigLabel: 'Conflicting' },
              { clinsig: -2, clinsigLabel: 'Benign' },
              { clinsig: -1, clinsigLabel: 'Likely benign' },
              { clinsig: 0, clinsigLabel: 'Uncertain significance' },
              { clinsig: 1, clinsigLabel: 'Likely pathogenic' },
              { clinsig: 2, clinsigLabel: 'Pathogenic' },
              { clinsig: 3, clinsigLabel: 'Gene' }
            ]
          },
          key: 'clinsig',
          fields: ['clinsigLabel']
        }
      }
    ],
    mark: { type: 'circle', opacity: 0.8 },
    encoding: {
      x: {
        field: 'pos',
        type: 'quantitative',
        scale: { domain: paddedMinMax.value },
        axis: { grid: false, zindex: 1000 },
        title: null
      },
      y: {
        field: 'clinsigLabel',
        type: 'nominal',
        scale: {
          domain: [
            'Gene',
            'Pathogenic',
            'Likely pathogenic',
            'Uncertain significance',
            'Likely benign',
            'Benign'
          ]
        },
        axis: { grid: false },
        title: null
      },
      color: {
        field: 'clinsigLabel',
        type: 'nominal',
        scale: {
          domain: [
            'Gene',
            'Pathogenic',
            'Likely pathogenic',
            'Uncertain significance',
            'Likely benign',
            'Benign'
          ],
          range: ['gray', 'darkred', 'orange', 'yellow', 'green', 'darkgreen']
        },
        legend: null
      },
      size: { value: 100 }
    }
  },
  {
    description: 'lollipop sticks',
    transform: [
      {
        lookup: 'clinsig',
        from: {
          data: {
            values: [
              {
                clinsig: -5,
                clinsigBaseline: 'Uncertain significance',
                clinsigLabel: 'other'
              },
              {
                clinsig: -4,
                clinsigBaseline: 'Uncertain significance',
                clinsigLabel: 'Not provided'
              },
              {
                clinsig: -3,
                clinsigBaseline: 'Uncertain significance',
                clinsigLabel: 'Conflicting'
              },
              {
                clinsig: -2,
                clinsigBaseline: 'Uncertain significance',
                clinsigLabel: 'Benign'
              },
              {
                clinsig: -1,
                clinsigBaseline: 'Uncertain significance',
                clinsigLabel: 'Likely benign'
              },
              {
                clinsig: 0,
                clinsigBaseline: 'Uncertain significance',
                clinsigLabel: 'Uncertain significance'
              },
              {
                clinsig: 1,
                clinsigBaseline: 'Uncertain significance',
                clinsigLabel: 'Likely pathogenic'
              },
              {
                clinsig: 2,
                clinsigBaseline: 'Uncertain significance',
                clinsigLabel: 'Pathogenic'
              },
              {
                clinsig: 3,
                clinsigBaseline: 'Uncertain significance',
                clinsigLabel: 'Gene'
              }
            ]
          },
          key: 'clinsig',
          fields: ['clinsigBaseline', 'clinsigLabel']
        }
      }
    ],
    mark: { type: 'rule', opacity: 0.8 },
    encoding: {
      x: {
        field: 'pos',
        type: 'quantitative',
        scale: { domain: paddedMinMax.value },
        axis: { grid: false },
        title: null
      },
      y: {
        field: 'clinsigLabel',
        type: 'nominal',
        scale: {
          domain: [
            'Gene',
            'Pathogenic',
            'Likely pathogenic',
            'Uncertain significance',
            'Likely benign',
            'Benign'
          ]
        },
        axis: { grid: false, zindex: 1000 },
        title: null
      },
      y2: { field: 'clinsigBaseline' },
      color: {
        field: 'clinsigLabel',
        type: 'nominal',
        scale: {
          domain: [
            'Gene',
            'Pathogenic',
            'Likely pathogenic',
            'Uncertain significance',
            'Likely benign',
            'Benign'
          ],
          range: ['gray', 'darkred', 'orange', 'yellow', 'green', 'darkgreen']
        },
        legend: null
      },
      size: { value: 1 }
    }
  },
  {
    description: 'gene - line',
    data: { values: [{ pos: minMax.value[0] }, { pos: minMax.value[1] }] },
    mark: { type: 'line', stroke: 'black', size: 1, opacity: 0.5 },
    encoding: {
      x: {
        field: 'pos',
        type: 'quantitative',
        scale: { domain: paddedMinMax.value },
        axis: { grid: false },
        title: null
      },
      y: { datum: 'Gene' }
    }
  },
  {
    description: 'gene - exons',
    data: {
      values: exons.value
    },
    mark: {
      type: 'rect',
      stroke: 'black',
      height: 10,
      fill: 'black',
      opacity: 0.5
    },
    encoding: {
      x: {
        field: 'start',
        type: 'quantitative',
        scale: { domain: paddedMinMax.value },
        axis: { grid: false },
        title: null
      },
      x2: { field: 'stop' },
      y: { datum: 'Gene' }
    }
  }
]
</script>

<template>
  <v-card>
    <v-card-title> ClinVar Variation </v-card-title>
    <v-divider />
    <figure class="figure border rounded pl-2 pt-2 mr-3 w-100 col">
      <figcaption class="figure-caption text-center">
        Variantion landscape for gene {{ props?.hgnc }}
      </figcaption>
      <VegaPlot
        :data-values="vegaData"
        :encoding="vegaEncoding"
        :layer="vegaLayer"
        :height="300"
        renderer="canvas"
      />
    </figure>
  </v-card>
</template>
