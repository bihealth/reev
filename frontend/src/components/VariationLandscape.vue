<script setup lang="ts">
import { computed } from 'vue'

import VegaPlot from '@/components/VegaPlot.vue'

export interface Props {
  /** Gene information from annonars. */
  clinvar: any
  /** Transctipts information. */
  transcripts: any
  /** The genome release. */
  genomeRelease: string
  /** The gene HGNC symbol. */
  hgnc: string
}

const props = withDefaults(defineProps<Props>(), {
  clinvar: null,
  transcripts: null,
  genomeRelease: 'grch37',
  hgnc: ''
})

const clinvarSignificanceMapping: Record<number, number> = {
  0: 0,
  1: -3,
  2: -2,
  3: -1,
  4: 1,
  5: 2
}

interface ClinvarVariant {
  chrom: string
  pos: string
  reference: string
  alternative: string
  rcv: string
  clinsig: number
  review_status: number
}

const convertClinvarSignificance = (input: number): number => {
  if (input in clinvarSignificanceMapping) {
    return clinvarSignificanceMapping[input]
  } else {
    return -4
  }
}

const minMax = computed(() => {
  if (!props.clinvar) {
    return []
  }
  let min = null
  let max = null
  for (const item of props.clinvar.variants ?? []) {
    if (item.genome_release.toLowerCase() == props.genomeRelease) {
      // Go through all item.variants and find the min and max pos.
      for (const variant of item.variants) {
        if (variant.pos < min || min == null) {
          min = variant.pos
        }
        if (variant.pos > max || max == null) {
          max = variant.pos
        }
      }
    }
  }
  return [min, max]
})

const exons = computed(() => {
  if (!props.transcripts) {
    return []
  }
  const exons = []
  for (const transcript of props.transcripts.transcripts) {
    for (const exon of transcript.exons) {
      exons.push({
        start: exon.start,
        stop: exon.end
      })
    }
  }
  return exons
})

const vegaData = computed(() => {
  if (!props.clinvar) {
    return []
  }
  let clinvarInfo = []
  for (const item of props.clinvar.variants ?? []) {
    if (item.genome_release.toLowerCase() == props.genomeRelease) {
      clinvarInfo = item.variants
    }
  }

  return clinvarInfo.map((variant: ClinvarVariant) => ({
    pos: variant.pos,
    clinsig: convertClinvarSignificance(variant.clinsig)
  }))
})

const vegaEncoding = {}

const vegaLayer = [
  {
    description: 'gray baseline',
    data: { values: [{}] },
    mark: { type: 'rule', stroke: 'lightgray', size: 3 },
    encoding: { y: { datum: 'Uncertain significance' } }
  },
  {
    description: 'lollipop heads',
    transform: [
      {
        lookup: 'clinsig',
        from: {
          data: {
            values: [
              { clinsig: -5, clinsigLabel: 'other' },
              { clinsig: -4, clinsigLabel: 'Not provided' },
              { clinsig: -3, clinsigLabel: 'Conflicting' },
              { clinsig: -2, clinsigLabel: 'Benign' },
              { clinsig: -1, clinsigLabel: 'Likely benign' },
              { clinsig: 0, clinsigLabel: 'Uncertain significance' },
              { clinsig: 1, clinsigLabel: 'Likely pathogenic' },
              { clinsig: 2, clinsigLabel: 'Pathogenic' },
              { clinsig: 3, clinsigLabel: 'Gene' }
            ]
          },
          key: 'clinsig',
          fields: ['clinsigLabel']
        }
      }
    ],
    mark: { type: 'circle', opacity: 0.8 },
    encoding: {
      x: {
        field: 'pos',
        type: 'quantitative',
        scale: { domain: minMax.value },
        axis: { grid: false, zindex: 1000 },
        title: null
      },
      y: {
        field: 'clinsigLabel',
        type: 'nominal',
        scale: {
          domain: [
            'Gene',
            'Pathogenic',
            'Likely pathogenic',
            'Uncertain significance',
            'Likely benign',
            'Benign'
          ]
        },
        axis: { grid: false },
        title: null
      },
      color: {
        field: 'clinsigLabel',
        type: 'nominal',
        scale: {
          domain: [
            'Gene',
            'Pathogenic',
            'Likely pathogenic',
            'Uncertain significance',
            'Likely benign',
            'Benign'
          ],
          range: ['gray', 'darkred', 'orange', 'yellow', 'green', 'darkgreen']
        },
        legend: null
      },
      size: { value: 100 }
    }
  },
  {
    description: 'lollipop sticks',
    transform: [
      {
        lookup: 'clinsig',
        from: {
          data: {
            values: [
              {
                clinsig: -5,
                clinsigBaseline: 'Uncertain significance',
                clinsigLabel: 'other'
              },
              {
                clinsig: -4,
                clinsigBaseline: 'Uncertain significance',
                clinsigLabel: 'Not provided'
              },
              {
                clinsig: -3,
                clinsigBaseline: 'Uncertain significance',
                clinsigLabel: 'Conflicting'
              },
              {
                clinsig: -2,
                clinsigBaseline: 'Uncertain significance',
                clinsigLabel: 'Benign'
              },
              {
                clinsig: -1,
                clinsigBaseline: 'Uncertain significance',
                clinsigLabel: 'Likely benign'
              },
              {
                clinsig: 0,
                clinsigBaseline: 'Uncertain significance',
                clinsigLabel: 'Uncertain significance'
              },
              {
                clinsig: 1,
                clinsigBaseline: 'Uncertain significance',
                clinsigLabel: 'Likely pathogenic'
              },
              {
                clinsig: 2,
                clinsigBaseline: 'Uncertain significance',
                clinsigLabel: 'Pathogenic'
              },
              {
                clinsig: 3,
                clinsigBaseline: 'Uncertain significance',
                clinsigLabel: 'Gene'
              }
            ]
          },
          key: 'clinsig',
          fields: ['clinsigBaseline', 'clinsigLabel']
        }
      }
    ],
    mark: { type: 'rule', opacity: 0.8 },
    encoding: {
      x: {
        field: 'pos',
        type: 'quantitative',
        scale: { domain: minMax.value },
        axis: { grid: false },
        title: null
      },
      y: {
        field: 'clinsigLabel',
        type: 'nominal',
        scale: {
          domain: [
            'Gene',
            'Pathogenic',
            'Likely pathogenic',
            'Uncertain significance',
            'Likely benign',
            'Benign'
          ]
        },
        axis: { grid: false, zindex: 1000 },
        title: null
      },
      y2: { field: 'clinsigBaseline' },
      color: {
        field: 'clinsigLabel',
        type: 'nominal',
        scale: {
          domain: [
            'Gene',
            'Pathogenic',
            'Likely pathogenic',
            'Uncertain significance',
            'Likely benign',
            'Benign'
          ],
          range: ['gray', 'darkred', 'orange', 'yellow', 'green', 'darkgreen']
        },
        legend: null
      },
      size: { value: 1 }
    }
  },
  {
    description: 'gene - line',
    data: { values: [{ pos: 41196312 }, { pos: 41277381 }] },
    mark: { type: 'line', stroke: 'black', size: 1, opacity: 0.5 },
    encoding: {
      x: {
        field: 'pos',
        type: 'quantitative',
        scale: { domain: minMax.value },
        axis: { grid: false },
        title: null
      },
      y: { datum: 'Gene' }
    }
  },
  {
    description: 'gene - exons',
    data: {
      values: exons.value
    },
    mark: {
      type: 'rect',
      stroke: 'black',
      height: 10,
      fill: 'black',
      opacity: 0.5
    },
    encoding: {
      x: {
        field: 'start',
        type: 'quantitative',
        scale: { domain: minMax.value },
        axis: { grid: false },
        title: null
      },
      x2: { field: 'stop' },
      y: { datum: 'Gene' }
    }
  }
]
</script>

<template>
  <figure class="figure border rounded pl-2 pt-2 mr-3 w-100 col">
    <figcaption class="figure-caption text-center">
      Variation Landscape of {{ props.hgnc }}
    </figcaption>
    <div style="width: 1100px; height: 350px; overflow: none">
      <VegaPlot
        :data-values="vegaData"
        :encoding="vegaEncoding"
        :layer="vegaLayer"
        :width="1000"
        :height="300"
        renderer="canvas"
      />
    </div>
  </figure>
</template>
