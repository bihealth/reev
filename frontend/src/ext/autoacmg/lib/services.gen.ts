// This file is auto-generated by @hey-api/openapi-ts

import { createClient, createConfig, type Options } from '@hey-api/client-fetch';
import type { PredictSeqvarApiV1PredictSeqvarGetData, PredictSeqvarApiV1PredictSeqvarGetError, PredictSeqvarApiV1PredictSeqvarGetResponse, PredictStrucvarApiV1PredictStrucvarGetData, PredictStrucvarApiV1PredictStrucvarGetError, PredictStrucvarApiV1PredictStrucvarGetResponse, ResolveVariantApiV1ResolveGetData, ResolveVariantApiV1ResolveGetError, ResolveVariantApiV1ResolveGetResponse } from './types.gen';

export const client = createClient(createConfig());

/**
 * Predict Seqvar
 * Predict the ACMG classification of a sequence variant.
 *
 * Args:
 * variant_name (str): The name or identifier of the sequence variant.
 * genome_release (str): The genome release version.
 *
 * Returns:
 * SeqVarPredictionResponse: The predicted ACMG classification.
 */
export const predictSeqvarApiV1PredictSeqvarGet = <ThrowOnError extends boolean = false>(options: Options<PredictSeqvarApiV1PredictSeqvarGetData, ThrowOnError>) => { return (options?.client ?? client).get<PredictSeqvarApiV1PredictSeqvarGetResponse, PredictSeqvarApiV1PredictSeqvarGetError, ThrowOnError>({
    ...options,
    url: '/api/v1/predict/seqvar'
}); };

/**
 * Predict Strucvar
 * Predict the ACMG classification of a structural variant.
 *
 * Args:
 * variant_name (str): The name or identifier of the structural variant.
 * genome_release (str): The genome release version.
 * duplication_tandem (bool): The duplication is in tandem and disrupts reading frame and undergoes NMD.
 *
 * Returns:
 * StrucVarPredictionResponse: The predicted ACMG classification.
 */
export const predictStrucvarApiV1PredictStrucvarGet = <ThrowOnError extends boolean = false>(options: Options<PredictStrucvarApiV1PredictStrucvarGetData, ThrowOnError>) => { return (options?.client ?? client).get<PredictStrucvarApiV1PredictStrucvarGetResponse, PredictStrucvarApiV1PredictStrucvarGetError, ThrowOnError>({
    ...options,
    url: '/api/v1/predict/strucvar'
}); };

/**
 * Resolve Variant
 * Resolve a variant to a sequence variant or structural variant.
 *
 * Args:
 * variant_name (str): The name or identifier of the variant.
 * genome_release (str): The genome release version.
 *
 * Returns:
 * VariantResolveResponse: The resolved variant.
 */
export const resolveVariantApiV1ResolveGet = <ThrowOnError extends boolean = false>(options: Options<ResolveVariantApiV1ResolveGetData, ThrowOnError>) => { return (options?.client ?? client).get<ResolveVariantApiV1ResolveGetResponse, ResolveVariantApiV1ResolveGetError, ThrowOnError>({
    ...options,
    url: '/api/v1/resolve'
}); };