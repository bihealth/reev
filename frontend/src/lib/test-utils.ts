import { type TestingPinia, createTestingPinia } from '@pinia/testing'
import { type VueWrapper, flushPromises, mount } from '@vue/test-utils'
import { vi } from 'vitest'
import { h } from 'vue'
import { type Router, createRouter, createWebHistory } from 'vue-router'
import { createVuetify } from 'vuetify'
import { md3 } from 'vuetify/blueprints'
import * as components from 'vuetify/components'
import * as directives from 'vuetify/directives'

import { routes } from '@/router'

/**
 * Return value for `setupMountedComponents`.
 */
export interface MountedComponents {
  /** The testing pinia instance. */
  pinia: TestingPinia
  /** The `VueWrapper` instance */
  wrapper: VueWrapper
  /** The router instance */
  router: Router
}

/**
 * Creates the `VueWrapper` for the given component with mounting as template.
 *
 * Having a shared utility wrapper helps us to reduce the code in the individual tests.
 *
 * Effectively, this will:
 *
 * - create an appropriate Vuetify instance
 * - create a router with all real routes but `router.push` mocked
 * - create a testing pinia with the given initial state
 *
 * @returns
 */
export const setupMountedComponents = async (
  componentOptions: {
    /** The component itself */
    component: any
  },
  options?: {
    /** Initial Store instances */
    initialStoreState?: any
    /** Props to pass to the component */
    props?: any
    /** Query to pass to the router */
    query?: any
    /** A custom pinia instance to use. Use this option only if you need to mock a store getter
     * or action.
     */
    pinia?: TestingPinia
  }
): Promise<MountedComponents> => {
  // Create new vuetify instance.
  const vuetify = createVuetify({
    blueprint: md3,
    components: {
      ...components
    },
    directives
  })

  // Create a `Router` instance that will be used in the tests and mock `push()`.
  const router = createRouter({
    history: createWebHistory(),
    routes: routes
  })
  router.push = vi.fn()
  router.replace = vi.fn()

  // Create a testing pinia with the initial data.
  const pinia = createTestingPinia({
    createSpy: vi.fn,
    initialState: options?.initialStoreState ?? {}
  })

  // Setup the component mapping that is to be tested.
  const knownComponents: { [key: string]: any } = {}
  knownComponents[componentOptions.component.__name] = componentOptions.component

  // Setup mounting option of component
  const componentMount = h(
    components.VApp,
    // props
    {},
    // children
    () => [h(componentOptions.component, options?.props ?? {})]
  )

  const wrapper = mount(componentMount, {
    query: options?.query,
    global: {
      plugins: [vuetify, router, options?.pinia ?? pinia],
      components: knownComponents
    }
  })

  await flushPromises()
  await vi.dynamicImportSettled()

  return { pinia, wrapper, router }
}
