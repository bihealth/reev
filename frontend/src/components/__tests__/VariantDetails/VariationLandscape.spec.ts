import { describe, expect, it } from 'vitest'

import VariationLandscape from '@/components/VariantDetails/VariationLandscape.vue'
import { setupMountedComponents } from '@/lib/test-utils'

const smallVariantInfo = {
  release: 'grch37',
  chromosome: 'chr17',
  start: '43044295',
  end: '43044295',
  reference: 'G',
  alternative: 'A',
  hgnc_id: 'HGNC:1100'
}

describe.concurrent('VariationLandscape', async () => {
  it('renders the VariationLandscape plot', async () => {
    const { wrapper } = setupMountedComponents(
      { component: VariationLandscape, template: false },
      {
        props: {
          clinvar: null,
          genomeRelease: null, 
          geneSymbol: null,
        }
      }
    )
    expect(wrapper.text()).toContain('Query Beacon -----|>')
    const refreshButton = wrapper.find('.mdi-refresh')
    const info = wrapper.find('.text-muted')
    expect(refreshButton.exists()).toBe(true)
    expect(info.exists()).toBe(true)
  })
})
