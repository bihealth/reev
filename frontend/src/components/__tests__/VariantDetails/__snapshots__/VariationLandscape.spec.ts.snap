// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`VariationLandscape > renders the VariationLandscape plot 1`] = `
"<figure class=\\"figure border rounded pl-2 pt-2 mr-3 w-100 col\\">
  <figcaption class=\\"figure-caption text-center\\"> Variation landscape of HGNC:1100</figcaption>
  <div></div>
</figure>"
`;
